api/client/client.go

```diff
@@ -0,0 +1,207 @@
+func (c *Client) ModelCreate(storeID, modelName string, items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.ModelCreateRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		Values:    values,
+	}
+
+	resp, err := c.client.ModelCreate(c.ctx, req)
+	if err != nil {
+		return err
+	}
+
+	for i, entity := range resp.GetEntities() {
+		err := json.Unmarshal([]byte(entity), items[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ModelSave saves existing instances
+func (c *Client) ModelSave(storeID, modelName string, items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.ModelSaveRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		Values:    values,
+	}
+	_, err = c.client.ModelSave(c.ctx, req)
+	return err
+}
+
+// ModelDelete deletes data
+func (c *Client) ModelDelete(storeID, modelName string, entityIDs ...string) error {
+	req := &pb.ModelDeleteRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	_, err := c.client.ModelDelete(c.ctx, req)
+	return err
+}
+
+// ModelHas checks if the specified entities exist
+func (c *Client) ModelHas(storeID, modelName string, entityIDs ...string) (bool, error) {
+	req := &pb.ModelHasRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	resp, err := c.client.ModelHas(c.ctx, req)
+	return resp.GetExists(), err
+}
+
+// ModelFind finds records by query
+func (c *Client) ModelFind(storeID, modelName string) (string, error) {
+	// TODO: implement query object
+	req := &pb.ModelFindRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+	}
+	resp, err := c.client.ModelFind(c.ctx, req)
+	return resp.GetEntity(), err
+}
+
+// ModelFindByID finds a record by id
+func (c *Client) ModelFindByID(storeID, modelName, entityID string, entity interface{}) error {
+	req := &pb.ModelFindByIDRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityID:  entityID,
+	}
+	resp, err := c.client.ModelFindByID(c.ctx, req)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(resp.GetEntity()), entity)
+	return err
+}
+
+// ReadTransaction returns a read transaction that can be started and used and ended
+func (c *Client) ReadTransaction(storeID, modelName string) (*ReadTransaction, error) {
+	client, err := c.client.ReadTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &ReadTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// Listen provides an update whenever the specified model is updated
+func (c *Client) Listen(storeID, modelName, entityID string) (chan interface{}, error) {
+	req := &pb.ListenRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityID:  entityID,
+	}
+	stream, err := c.client.Listen(c.ctx, req)
+	if err != nil {
+		return nil, err
+	}
+	channel := make(chan interface{})
```

Comment:
@sanderpick @jsign @carsonfarmer I don't really know how to deal with this situation since there are no generics in Go. This is the api where the user can subscribe to updates on a entity instance. You can see here that I'm making a `channel` of some unknown type (it could also be of `map[string]interface{}`), and then below I'm unmarshalling the json string into a `map[string]interface{}`. Not a great api. How can this be improved?

Reply:
slight improvement passing in a dummy instance, but not sure